pyevmasm/evmasm.py

```diff
@@ -1090,6 +596,293 @@ def assemble_hex(asmcode, pc=0, fork=DEFAULT_FORK):
+        Instruction(0x0, 'STOP', 0, 0, 0, 0, 'Halts execution.', None, 0)
+        >>> instruction_table['STOP']
+        Instruction(0x0, 'STOP', 0, 0, 0, 0, 'Halts execution.', None, 0)
+        >>> i = instruction_table.__iter__()
+        >>> i.__next__()
+        Instruction(0x0, 'STOP', 0, 0, 0, 0, 'Halts execution.', None, 0)
+        >>> i.__next__()
+        Instruction(0x1, 'ADD', 0, 2, 1, 3, 'Addition operation.', None, 0)
+        >>> i.__next__()
+        Instruction(0x2, 'MUL', 0, 2, 1, 5, 'Multiplication operation.', None, 0)
+        >>> i.__next__()
+        Instruction(0x3, 'SUB', 0, 2, 1, 3, 'Subtraction operation.', None, 0)
+
+    """
+    __slots__ = ('_previous_fork', '_ilist', '_cache', '_name_to_opcode')
+    def __init__(self, *args, **kwargs):
+        self._previous_fork = kwargs.get('previous_fork', None)
+        self._ilist = dict(args[0])
+        self._cache = {}
+        self._name_to_opcode = None
+
+    def _search_by_opcode(self, k):
+        try:
+            return (k, *self._ilist[k])
+        except KeyError:
+            if self._previous_fork is not None:
+                return self._previous_fork._search_by_opcode(k)
+        raise KeyError(k)
+
+    def _search_by_name(self, k):
+        if self._name_to_opcode is None:
+            self._name_to_opcode = {}
+            for opcode, (name, operand_size, pushes, pops, gas, description) in self._ilist.items():
+                if name == 'PUSH':
+                    long_name = 'PUSH%d' % operand_size
+                elif name == 'DUP':
+                    long_name = 'DUP%d' % pops
+                elif name == 'SWAP':
+                    long_name = 'SWAP%d' % (pops - 1)
+                elif name == 'LOG':
+                    long_name ='LOG%d' % (pops - 2)
+                else:
+                    long_name = name
+                self._name_to_opcode[long_name] = opcode
+
+        try:
+            return self._search_by_opcode(self._name_to_opcode[k])
+        except KeyError:
+            if self._previous_fork is not None:
+                return self._previous_fork._search_by_name(k)
+        raise KeyError(k)
+
+    def _search(self, k):
+        if k in self._cache:
+            return self._cache[k]
+        try:
+            value = self._search_by_opcode(k)
+        except KeyError:
+            value = self._search_by_name(k)
+        self._cache[k] = value
+        return value
+
+    def __getitem__(self, k):
+        return Instruction(*self._search(k))
+
+    def get(self, k, default=None):
+        try:
+            return Instruction(*self._search(k))
+        except KeyError:
+            return default
+
+    def __contains__(self, k):
+        if k not in self._ilist:
+            if self._previous_fork is not None:
+                return k in self._previous_fork
+
+    def keys(self):
+        keys = ()
+        if self._previous_fork is not None:
+            keys = self._previous_fork.keys()
```

Comment:
if we keep it, it needs to be `keys = tuple(...keys())` otherwise it will crash in py3:
```
In [6]: (1,2) + {}.keys()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-6-a1e90b156e6e> in <module>
----> 1 (1,2) + {}.keys()

TypeError: can only concatenate tuple (not "dict_keys") to tuple```